### PR TITLE
[test-only] new lock tests

### PR DESCRIPTION
**Pull request by ScharfViktor**
related #7561

We have lock test which we moved from core repo. Out of 98 scenarios related to locking, only 6 of them work. 
and the existing tests can't: 
- set a timeout
- work in project space
- use space dav
- no tests that check if the lock expires

I created first PR where: 
- [x] user can create lock using different path
- [x] user can create lock using fileId - Collabora and onlyOfiice locks file using this way
- [x] user can set timeout
- [x] space members with edit permissions can create file lock  
- [x] share receiver  with edit permissions can create file lock - works only using fileId

found issue: 
- https://github.com/owncloud/ocis/issues/7599
- https://github.com/owncloud/ocis/issues/7600

next test, next PRs: 
- unlock file
- check if the lock expires

### CODE REVIEW

File: tests/acceptance/features/apiSpaces/lockFiles.feature [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+      | d:lockdiscovery/d:activelock/d:timeout               | Infinity     |
+      | d:lockdiscovery/d:activelock/oc:ownername            | Alice Hansen |
+    Examples:
+      | dav-path-version |
+      | old              |
+      | new              |
+      | spaces           |
+
+
+  Scenario Outline: locking a file with a set timeout
```
Comment (saw-jan) [inline tests/acceptance/features/apiSpaces/lockFiles.feature:33]:
```suggestion
  Scenario Outline: locking a file with a timeout
```

File: tests/acceptance/features/apiSpaces/lockFiles.feature [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+  I want to lock files
+
+  Background:
+    Given these users have been created with default attributes and without skeleton files:
+      | username |
+      | Alice    |
+      | Brian    |
+
+
+  Scenario Outline: locking a file
```
Comment (saw-jan) [inline tests/acceptance/features/apiSpaces/lockFiles.feature:12]:
```suggestion
  Scenario Outline: lock a file
```

File: tests/acceptance/features/apiSpaces/lockFiles.feature [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+      | d:lockdiscovery/d:activelock/d:timeout               | Second-5000  |
+      | d:lockdiscovery/d:activelock/oc:ownername            | Alice Hansen |
+    Examples:
+      | dav-path-version |
+      | old              |
+      | new              |
+      | spaces           |
+
+
+  Scenario Outline: locking a file by file-id
```
Comment (saw-jan) [inline tests/acceptance/features/apiSpaces/lockFiles.feature:55]:
```suggestion
  Scenario Outline: lock a file by file-id
```

File: tests/acceptance/features/apiSpaces/lockFiles.feature [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+      | lockscope | exclusive |
+    Then the HTTP status code should be "423"
+    Examples:
+      | dav-path-version |
+      | old              |
+      | new              |
+      | spaces           |
+
+
+  Scenario Outline: locking a file in the project space
```
Comment (saw-jan) [inline tests/acceptance/features/apiSpaces/lockFiles.feature:91]:
```suggestion
  Scenario Outline: lock a file in the project space
```

File: tests/acceptance/features/apiSpaces/lockFiles.feature [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+      | d:lockdiscovery/d:activelock/d:depth                 | Infinity     |
+      | d:lockdiscovery/d:activelock/d:timeout               | Second-3600  |
+      | d:lockdiscovery/d:activelock/oc:ownername            | Brian Murphy |
+    Examples:
+      | role    |
+      | manager |
+      | editor  |
+
+
+  Scenario Outline: locking a file in the project space by file-id
```
Comment (saw-jan) [inline tests/acceptance/features/apiSpaces/lockFiles.feature:117]:
```suggestion
  Scenario Outline: lock a file in the project space by file-id
```

File: tests/acceptance/features/apiSpaces/lockFiles.feature [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+      | role      | viewer |
+    When user "Brian" locks file "textfile.txt" inside the space "Project" using the WebDAV API setting the following properties
+      | lockscope | exclusive |
+    Then the HTTP status code should be "403"
+    When user "Brian" locks file using file-id path "/dav/spaces/<<FILEID>>" using the WebDAV API setting the following properties
+      | lockscope | exclusive |
+    Then the HTTP status code should be "403"
+
+
+  Scenario Outline: locking a file in the shares
```
Comment (saw-jan) [inline tests/acceptance/features/apiSpaces/lockFiles.feature:161]:
```suggestion
  Scenario Outline: lock a file in the shares
```

File: tests/acceptance/features/apiSpaces/lockFiles.feature [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+    Given using <dav-path-version> DAV path
+    And user "Alice" has uploaded a file inside space "Alice Hansen" with content "some content" to "textfile.txt"
+    When user "Alice" creates a share inside of space "Alice Hansen" with settings:
```
Comment (saw-jan) [inline tests/acceptance/features/apiSpaces/lockFiles.feature:164]: this step can be done as Given step
```suggestion
    Given user "Alice" creates a share inside of space "Alice Hansen" with settings:
```

File: tests/acceptance/features/apiSpaces/lockFiles.feature [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+      | d:lockdiscovery/d:activelock/d:lockscope/d:exclusive |              |
+      | d:lockdiscovery/d:activelock/oc:ownername            | Brian Murphy |
+    Examples:
+      | dav-path-version |
+      | old              |
+      | new              |
+      | spaces           |
+
+
+  Scenario Outline: locking a file in the shares using file-id
```
Comment (saw-jan) [inline tests/acceptance/features/apiSpaces/lockFiles.feature:184]:
```suggestion
  Scenario Outline: lock a file in the shares using file-id
```

File: tests/acceptance/features/bootstrap/WebDavLockingContext.php [hunk 9]
```diff
@@ -113,7 +116,121 @@ private function lockFile(
...
+	 * @param boolean $expectToSucceed
+	 *
+	 * @return void
+	 */
+	private function lockFileInProjectSpace(
+		string $user,
+		string $file,
+		string $space,
+		TableNode $properties,
+		bool $expectToSucceed = true
```
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/WebDavLockingContext.php:133]:
```suggestion
	private function lockFileInSpace(
	        string $fullUrl,
		string $user,
		string $file,
		TableNode $properties,
		bool $expectToSucceed = true
```

File: tests/acceptance/features/bootstrap/WebDavLockingContext.php [hunk 10]
```diff
@@ -113,7 +116,121 @@ private function lockFile(
...
+	/**
+	 *
+	 * @param string $user
+	 * @param string $filePath
+	 * @param TableNode $properties table with no heading with | property | value |
+	 * @param boolean $expectToSucceed
+	 *
+	 * @return void
+	 */
+	private function lockFileUsingFileId(
```
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/WebDavLockingContext.php:185]:
```suggestion
```

File: tests/acceptance/features/bootstrap/WebDavLockingContext.php [hunk 11]
```diff
@@ -127,7 +244,36 @@ public function lockFileUsingWebDavAPI(string $user, string $file, TableNode $pr
...
+	 *
+	 * @param string $user
+	 * @param string $file
+	 * @param string $space
+	 * @param TableNode $properties table with no heading with | property | value |
+	 *
+	 * @return void
+	 */
+	public function lockFileInProjectSpaceUsingWebDavAPI(string $user, string $file, string $space, TableNode $properties) {
+		$response = $this->lockFileInProjectSpace($user, $file, $space, $properties, false, false);
```
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/WebDavLockingContext.php:257]:
```suggestion
                $spaceId = $this->spacesContext->getSpaceIdByName($user, $space);
		$fullUrl = $this->featureContext->getBaseUrl() . '/dav/spaces/' . $spaceId . '/' . $file;
		$response = $this->lockFileInProjectSpace($fullUrl, $user, $file, $properties, false, false);
```

File: tests/acceptance/features/bootstrap/WebDavLockingContext.php [hunk 12]
```diff
@@ -127,7 +244,36 @@ public function lockFileUsingWebDavAPI(string $user, string $file, TableNode $pr
...
+	 * @When user :user locks file using file-id path :path using the WebDAV API setting the following properties
+	 *
+	 * @param string $user
+	 * @param string $filePath
+	 * @param TableNode $properties table with no heading with | property | value |
+	 *
+	 * @return void
+	 */
+	public function lockFileUsingFileIdUsingWebDavAPI(string $user, string $filePath, TableNode $properties) {
+		$response = $this->lockFileUsingFileId($user, $filePath, $properties, false, false);
```
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/WebDavLockingContext.php:271]:
```suggestion
                $fullUrl = $this->featureContext->getBaseUrl() . $filePath;
		$response = $this->lockFileUsingFileId($fullUrl, $user, $filePath, $properties, false, false);
```

**Review by saw-jan — COMMENTED** [review]
I see that `lockFileInProjectSpace` and `lockFileUsingFileId` almost identical except `fullUrl`. Maybe we can have single method for both actions

File: tests/acceptance/features/bootstrap/WebDavLockingContext.php [hunk 13]
```diff
@@ -742,5 +837,6 @@ public function before(BeforeScenarioScope $scope) {
 		// Get all the contexts you need in this context
 		$this->featureContext = $environment->getContext('FeatureContext');
 		$this->publicWebDavContext = $environment->getContext('PublicWebDavContext');
+		$this->spacesContext = $environment->getContext('SpacesContext');
```
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/WebDavLockingContext.php:840]: either we have to conditionally use `$this->spacesContext` or move spaces related steps to another context file (e.g. Spaces context itself). Reva will complain about it

File: tests/acceptance/features/bootstrap/WebDavLockingContext.php [hunk 14]
```diff
@@ -219,10 +159,10 @@ private function lockFileUsingFileId(
 		$responseXml = $this->featureContext->getResponseXml($response, __METHOD__);
 		$xmlPart = $responseXml->xpath("//d:locktoken/d:href");
 		if (isset($xmlPart[0])) {
-			$this->tokenOfLastLock[$user][$filePath] = (string) $xmlPart[0];
+			$this->tokenOfLastLock[$user] = (string) $xmlPart[0];
```
Comment (ScharfViktor) [inline tests/acceptance/features/bootstrap/WebDavLockingContext.php:162]: @saw-jan I'll solve storing lock token when I combine lockFileInSpace() and lockFile() and move tests from coreApiWebdavLocks to apiSpaces

File: tests/acceptance/features/bootstrap/WebDavLockingContext.php [hunk 15]
```diff
@@ -127,7 +184,39 @@ public function lockFileUsingWebDavAPI(string $user, string $file, TableNode $pr
...
+	 * @param string $file
+	 * @param string $space
+	 * @param TableNode $properties table with no heading with | property | value |
+	 *
+	 * @return void
+	 */
+	public function lockFileInProjectSpaceUsingWebDavAPI(string $user, string $file, string $space, TableNode $properties) {
+		$spaceId = $this->spacesContext->getSpaceIdByName($user, $space);
+		$fullUrl = $this->featureContext->getBaseUrl() . '/dav/spaces/' . $spaceId . '/' . $file;
+		$response = $this->lockFileInSpace($user, $fullUrl, $properties, false, false);
```
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/WebDavLockingContext.php:199]: can they be by default?
```suggestion
		$response = $this->lockFileInSpace($user, $fullUrl, $properties);
```

File: tests/acceptance/features/bootstrap/WebDavLockingContext.php [hunk 16]
```diff
@@ -127,7 +184,39 @@ public function lockFileUsingWebDavAPI(string $user, string $file, TableNode $pr
...
+	 *
+	 * @param string $user
+	 * @param string $filePath
+	 * @param TableNode $properties table with no heading with | property | value |
+	 *
+	 * @return void
+	 */
+	public function lockFileUsingFileIdUsingWebDavAPI(string $user, string $filePath, TableNode $properties) {
+		$fullUrl = $this->featureContext->getBaseUrl() . $filePath;
+		$response = $this->lockFileInSpace($user, $fullUrl, $properties, false, false);
```
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/WebDavLockingContext.php:214]:
```suggestion
		$response = $this->lockFileInSpace($user, $fullUrl, $properties);
```

**Review by saw-jan — APPROVED** [review]
LGTM :+1:

**Pull request merged**